File: backend/app/schemas/student_course.py
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime


class StudentCourseCreate(BaseModel):
    """建立學生選課"""
    student_id: str = Field(..., description="學生 ID")
    course_id: str = Field(..., description="課程 ID")

    model_config = {
        "json_schema_extra": {
            "examples": [{
                "student_id": "a1b2c3d4-e5f6-7890-abcd-ef1234567890",
                "course_id": "b2c3d4e5-f6a7-8901-bcde-f12345678901",
            }]
        }
    }


class StudentCourseResponse(BaseModel):
    """學生選課回應"""
    id: str
    student_id: str
    course_id: str
    course_code: Optional[str] = None
    course_name: Optional[str] = None
    student_name: Optional[str] = None
    enrolled_at: Optional[datetime] = None
    created_at: Optional[datetime] = None

    model_config = {
        "from_attributes": True,
        "json_schema_extra": {
            "examples": [{
                "id": "a1b2c3d4-e5f6-7890-abcd-ef1234567890",
                "student_id": "b2c3d4e5-f6a7-8901-bcde-f12345678901",
                "course_id": "c3d4e5f6-a7b8-9012-cdef-123456789012",
                "course_code": "ENG-A1",
                "course_name": "英語初級會話",
                "student_name": "王小明",
                "enrolled_at": "2026-03-01T10:00:00",
                "created_at": "2026-03-01T10:00:00",
            }]
        }
    }


class StudentCourseListResponse(BaseModel):
    """學生選課列表回應"""
    success: bool = True
    data: list[StudentCourseResponse] = []
    total: int = 0
    page: int = 1
    per_page: int = 20
    total_pages: int = 0

File: frontend/src/api/studentCourse.ts
import request from '@/utils/request';
import type { BaseResponse, DataResponse } from './response';

// ─── Types ─────────────────────────────────────────────────────────────────

export interface StudentCourseCreate {
  student_id: string;
  course_id: string;
}

export interface StudentCourseResponse {
  id: string;
  student_id: string;
  course_id: string;
  course_code: string | null;
  course_name: string | null;
  student_name: string | null;
  enrolled_at: string | null;
  created_at: string | null;
}

export interface StudentCourseListResponse {
  success: boolean;
  message?: string;
  error_code?: string | null;
  data: StudentCourseResponse[];
  total: number;
  page: number;
  per_page: number;
  total_pages: number;
}

export interface CourseOption {
  id: string;
  course_name: string;
  course_code?: string | null;
}

// ─── API Functions ──────────────────────────────────────────────────────────

/** 取得學生已選課程列表 */
export const getStudentCourses = (student_id: string) => {
  return request.get<any, StudentCourseListResponse>('/v1/student-courses', {
    params: { student_id, per_page: 100 },
  });
};

/** 取得某學生的已選課程（供合約明細 course 下拉用） */
export const getStudentCoursesByStudentId = (student_id: string) => {
  return request.get<any, StudentCourseListResponse>(`/v1/student-courses/by-student/${student_id}`);
};

/** 刪除已選課程 */
export const deleteStudentCourse = (enrollment_id: string) => {
  return request.delete<any, BaseResponse>(`/v1/student-courses/${enrollment_id}`);
};

/** 取得可選課程清單 (下拉選單用) */
export const getCourseOptions = () => {
  return request.get<any, DataResponse<CourseOption[]>>('/v1/student-courses/options/courses');
};

/** 新增學生選課 */
export const createStudentCourse = (data: StudentCourseCreate) => {
  return request.post<any, DataResponse<StudentCourseResponse>>('/v1/student-courses', data);
};
